### ISSUE TITLE
Yolov8 custom model failed to convert

### ISSUE THREAD

**Issue by jiangdizhao** [marcoslucianops/DeepStream-Yolo#295]
I trained a yolov8n model with 29 classes, according to the instructions, I generated .cfg and .wts files. Then, i  changed the parameter "num-detected-classes" to 29.    Finally,  when I tried to run deepstream-app -c deepstream_app_config.txt, it prompted the following error:
Number of unused weights left: 18446744073709540969
deepstream-app: yolo.cpp:426: NvDsInferStatus Yolo::buildYoloNetwork(std::vector<float>&, nvinfer1::INetworkDefinition&): Assertion `0' failed.
Discarded (core dumped)

I guess the parameter "num-detected-classes" were not used during model creation, instead it used the default number 80
Where can I modified this number in cpp or hpp files ?@marcoslucianops

**Comment by jiangdizhao** [comment]
all proccess is shown as follows:  



ERROR: ../nvdsinfer/nvdsinfer_model_builder.cpp:1484 Deserialize engine failed because file path: /home/rico/文档/DeepStream-Yolo/yolov8_v17.engine open error
0:00:00.435858438 17219 0x560df87dd070 WARN                 nvinfer gstnvinfer.cpp:635:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1889> [UID = 1]: deserialize engine from file :/home/rico/文档/DeepStream-Yolo/yolov8_v17.engine failed
0:00:00.435889985 17219 0x560df87dd070 WARN                 nvinfer gstnvinfer.cpp:635:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:1996> [UID = 1]: deserialize backend context from engine from file :/home/rico/文档/DeepStream-Yolo/yolov8_v17.engine failed, try rebuild
0:00:00.435896443 17219 0x560df87dd070 INFO                 nvinfer gstnvinfer.cpp:638:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1914> [UID = 1]: Trying to create engine from model files

Loading pre-trained weights
Loading weights of yolov8_v17 complete
Total weights read: 3041440
Building YOLO network

        Layer                         Input Shape         Output Shape        WeightPtr
(0)     conv_silu                     [3, 640, 640]       [16, 320, 320]      496
(1)     conv_silu                     [16, 320, 320]      [32, 160, 160]      5232
(2)     conv_silu                     [32, 160, 160]      [32, 160, 160]      6384
(3)     c2f_silu                      [32, 160, 160]      [48, 160, 160]      11120
(4)     conv_silu                     [48, 160, 160]      [32, 160, 160]      12784
(5)     conv_silu                     [32, 160, 160]      [64, 80, 80]        31472
(6)     conv_silu                     [64, 80, 80]        [64, 80, 80]        35824
(7)     c2f_silu                      [64, 80, 80]        [128, 80, 80]       73200
(8)     conv_silu                     [128, 80, 80]       [64, 80, 80]        81648
(9)     conv_silu                     [64, 80, 80]        [128, 40, 40]       155888
(10)    conv_silu                     [128, 40, 40]       [128, 40, 40]       172784
(11)    c2f_silu                      [128, 40, 40]       [256, 40, 40]       321264
(12)    conv_silu                     [256, 40, 40]       [128, 40, 40]       354544
(13)    conv_silu                     [128, 40, 40]       [256, 20, 20]       650480
(14)    conv_silu                     [256, 20, 20]       [256, 20, 20]       717040
(15)    c2f_silu                      [256, 20, 20]       [384, 20, 20]       1012976
(16)    conv_silu                     [384, 20, 20]       [256, 20, 20]       1112304
(17)    conv_silu                     [256, 20, 20]       [128, 20, 20]       1145584
(18)    maxpool                       [128, 20, 20]       [128, 20, 20]       -
(19)    maxpool                       [128, 20, 20]       [128, 20, 20]       -
(20)    maxpool                       [128, 20, 20]       [128, 20, 20]       -
(21)    route: 17, 18, 19, 20         -                   [512, 20, 20]       -
(22)    conv_silu                     [512, 20, 20]       [256, 20, 20]       1277680
(23)    upsample                      [256, 20, 20]       [256, 40, 40]       -
(24)    route: 23, 12                 -                   [384, 40, 40]       -
(25)    conv_silu                     [384, 40, 40]       [128, 40, 40]       1327344
(26)    c2f_silu                      [128, 40, 40]       [192, 40, 40]       1401584
(27)    conv_silu                     [192, 40, 40]       [128, 40, 40]       1426672
(28)    upsample                      [128, 40, 40]       [128, 80, 80]       -
(29)    route: 28, 8                  -                   [192, 80, 80]       -
(30)    conv_silu                     [192, 80, 80]       [64, 80, 80]        1439216
(31)    c2f_silu                      [64, 80, 80]        [96, 80, 80]        1457904
(32)    conv_silu                     [96, 80, 80]        [64, 80, 80]        1464304
(33)    conv_silu                     [64, 80, 80]        [64, 40, 40]        1501424
(34)    route: 33, 27                 -                   [192, 40, 40]       -
(35)    conv_silu                     [192, 40, 40]       [128, 40, 40]       1526512
(36)    c2f_silu                      [128, 40, 40]       [192, 40, 40]       1600752
(37)    conv_silu                     [192, 40, 40]       [128, 40, 40]       1625840
(38)    conv_silu                     [128, 40, 40]       [128, 20, 20]       1773808
(39)    route: 38, 22                 -                   [384, 20, 20]       -
(40)    conv_silu                     [384, 20, 20]       [256, 20, 20]       1873136
(41)    c2f_silu                      [256, 20, 20]       [384, 20, 20]       2169072
(42)    conv_silu                     [384, 20, 20]       [256, 20, 20]       2268400
(43)    route: 32                     -                   [64, 80, 80]        -
(44)    conv_silu                     [64, 80, 80]        [64, 80, 80]        2305520
(45)    conv_silu                     [64, 80, 80]        [64, 80, 80]        2342640
(46)    conv_linear                   [64, 80, 80]        [29, 80, 80]        2344525
(47)    route: 43                     -                   [64, 80, 80]        -
(48)    conv_silu                     [64, 80, 80]        [64, 80, 80]        2381645
(49)    conv_silu                     [64, 80, 80]        [64, 80, 80]        2418765
(50)    conv_linear                   [64, 80, 80]        [64, 80, 80]        2422925
(51)    route: 50, 46                 -                   [93, 80, 80]        -
(52)    shuffle: 51                   [93, 80, 80]        [93, 6400]          -
(53)    route: 37                     -                   [128, 40, 40]       -
(54)    conv_silu                     [128, 40, 40]       [64, 40, 40]        2496909
(55)    conv_silu                     [64, 40, 40]        [64, 40, 40]        2534029
(56)    conv_linear                   [64, 40, 40]        [29, 40, 40]        2535914
(57)    route: 53                     -                   [128, 40, 40]       -
(58)    conv_silu                     [128, 40, 40]       [64, 40, 40]        2609898
(59)    conv_silu                     [64, 40, 40]        [64, 40, 40]        2647018
(60)    conv_linear                   [64, 40, 40]        [64, 40, 40]        2651178
(61)    route: 60, 56                 -                   [93, 40, 40]        -
(62)    shuffle: 61                   [93, 40, 40]        [93, 1600]          -
(63)    route: 42                     -                   [256, 20, 20]       -
(64)    conv_silu                     [256, 20, 20]       [64, 20, 20]        2798890
(65)    conv_silu                     [64, 20, 20]        [64, 20, 20]        2836010
(66)    conv_linear                   [64, 20, 20]        [29, 20, 20]        2837895
(67)    route: 63                     -                   [256, 20, 20]       -
(68)    conv_silu                     [256, 20, 20]       [64, 20, 20]        2985607
(69)    conv_silu                     [64, 20, 20]        [64, 20, 20]        3022727
(70)    conv_linear                   [64, 20, 20]        [64, 20, 20]        3026887
(71)    route: 70, 66                 -                   [93, 20, 20]        -
(72)    shuffle: 71                   [93, 20, 20]        [93, 400]           -
(73)    route: 52, 62, 72             -                   [93, 8400]          -
(74)    detect_v8                     [93, 8400]          [8400, 33]          3052087

Number of unused weights left: 18446744073709540969
deepstream-app: yolo.cpp:426: NvDsInferStatus Yolo::buildYoloNetwork(std::vector<float>&, nvinfer1::INetworkDefinition&): Assertion `0' failed.

**Comment by marcoslucianops** [comment]
Can you send me the 1st epoch (don't need the last epoch) of your model to fix this error? I don't have a model with different number of classes here.

**Comment by jiangdizhao** [comment]
try this:
[epoch1.zip](https://github.com/marcoslucianops/DeepStream-Yolo/files/10575186/epoch1.zip)

**Comment by jiangdizhao** [comment]
This model contains 29 classes of objects @marcoslucianops

**Comment by marcoslucianops** [comment]
I updated the repo to fix the problem. Please convert your model again using the `gen_wts_yoloV8.py` file and test it.

**Comment by surbhi0826** [comment]
@marcoslucianops  now I am able to convert the model and build engine but predictions are way far from what was expected.

**Comment by marcoslucianops** [comment]
are there big difference when compared to pytorch? Can you send me one image comparing them?

**Comment by jiangdizhao** [comment]
Problem still remains:


Loading pre-trained weights
Loading weights of yolov8_v17 complete
Total weights read: 3041440
Building YOLO network

        Layer                         Input Shape         Output Shape        WeightPtr
(0)     conv_silu                     [3, 640, 640]       [16, 320, 320]      496
(1)     conv_silu                     [16, 320, 320]      [32, 160, 160]      5232
(2)     conv_silu                     [32, 160, 160]      [32, 160, 160]      6384
(3)     c2f_silu                      [32, 160, 160]      [48, 160, 160]      11120
(4)     conv_silu                     [48, 160, 160]      [32, 160, 160]      12784
(5)     conv_silu                     [32, 160, 160]      [64, 80, 80]        31472
(6)     conv_silu                     [64, 80, 80]        [64, 80, 80]        35824
(7)     c2f_silu                      [64, 80, 80]        [128, 80, 80]       73200
(8)     conv_silu                     [128, 80, 80]       [64, 80, 80]        81648
(9)     conv_silu                     [64, 80, 80]        [128, 40, 40]       155888
(10)    conv_silu                     [128, 40, 40]       [128, 40, 40]       172784
(11)    c2f_silu                      [128, 40, 40]       [256, 40, 40]       321264
(12)    conv_silu                     [256, 40, 40]       [128, 40, 40]       354544
(13)    conv_silu                     [128, 40, 40]       [256, 20, 20]       650480
(14)    conv_silu                     [256, 20, 20]       [256, 20, 20]       717040
(15)    c2f_silu                      [256, 20, 20]       [384, 20, 20]       1012976
(16)    conv_silu                     [384, 20, 20]       [256, 20, 20]       1112304
(17)    conv_silu                     [256, 20, 20]       [128, 20, 20]       1145584
(18)    maxpool                       [128, 20, 20]       [128, 20, 20]       -
(19)    maxpool                       [128, 20, 20]       [128, 20, 20]       -
(20)    maxpool                       [128, 20, 20]       [128, 20, 20]       -
(21)    route: 17, 18, 19, 20         -                   [512, 20, 20]       -
(22)    conv_silu                     [512, 20, 20]       [256, 20, 20]       1277680
(23)    upsample                      [256, 20, 20]       [256, 40, 40]       -
(24)    route: 23, 12                 -                   [384, 40, 40]       -
(25)    conv_silu                     [384, 40, 40]       [128, 40, 40]       1327344
(26)    c2f_silu                      [128, 40, 40]       [192, 40, 40]       1401584
(27)    conv_silu                     [192, 40, 40]       [128, 40, 40]       1426672
(28)    upsample                      [128, 40, 40]       [128, 80, 80]       -
(29)    route: 28, 8                  -                   [192, 80, 80]       -
(30)    conv_silu                     [192, 80, 80]       [64, 80, 80]        1439216
(31)    c2f_silu                      [64, 80, 80]        [96, 80, 80]        1457904
(32)    conv_silu                     [96, 80, 80]        [64, 80, 80]        1464304
(33)    conv_silu                     [64, 80, 80]        [64, 40, 40]        1501424
(34)    route: 33, 27                 -                   [192, 40, 40]       -
(35)    conv_silu                     [192, 40, 40]       [128, 40, 40]       1526512
(36)    c2f_silu                      [128, 40, 40]       [192, 40, 40]       1600752
(37)    conv_silu                     [192, 40, 40]       [128, 40, 40]       1625840
(38)    conv_silu                     [128, 40, 40]       [128, 20, 20]       1773808
(39)    route: 38, 22                 -                   [384, 20, 20]       -
(40)    conv_silu                     [384, 20, 20]       [256, 20, 20]       1873136
(41)    c2f_silu                      [256, 20, 20]       [384, 20, 20]       2169072
(42)    conv_silu                     [384, 20, 20]       [256, 20, 20]       2268400
(43)    route: 32                     -                   [64, 80, 80]        -
(44)    conv_silu                     [64, 80, 80]        [64, 80, 80]        2305520
(45)    conv_silu                     [64, 80, 80]        [64, 80, 80]        2342640
(46)    conv_linear                   [64, 80, 80]        [29, 80, 80]        2344525
(47)    route: 43                     -                   [64, 80, 80]        -
(48)    conv_silu                     [64, 80, 80]        [64, 80, 80]        2381645
(49)    conv_silu                     [64, 80, 80]        [64, 80, 80]        2418765
(50)    conv_linear                   [64, 80, 80]        [64, 80, 80]        2422925
(51)    route: 50, 46                 -                   [93, 80, 80]        -
(52)    shuffle                       [93, 80, 80]        [93, 6400]          -
(53)    route: 37                     -                   [128, 40, 40]       -
(54)    conv_silu                     [128, 40, 40]       [64, 40, 40]        2496909
(55)    conv_silu                     [64, 40, 40]        [64, 40, 40]        2534029
(56)    conv_linear                   [64, 40, 40]        [29, 40, 40]        2535914
(57)    route: 53                     -                   [128, 40, 40]       -
(58)    conv_silu                     [128, 40, 40]       [64, 40, 40]        2609898
(59)    conv_silu                     [64, 40, 40]        [64, 40, 40]        2647018
(60)    conv_linear                   [64, 40, 40]        [64, 40, 40]        2651178
(61)    route: 60, 56                 -                   [93, 40, 40]        -
(62)    shuffle                       [93, 40, 40]        [93, 1600]          -
(63)    route: 42                     -                   [256, 20, 20]       -
(64)    conv_silu                     [256, 20, 20]       [64, 20, 20]        2798890
(65)    conv_silu                     [64, 20, 20]        [64, 20, 20]        2836010
(66)    conv_linear                   [64, 20, 20]        [29, 20, 20]        2837895
(67)    route: 63                     -                   [256, 20, 20]       -
(68)    conv_silu                     [256, 20, 20]       [64, 20, 20]        2985607
(69)    conv_silu                     [64, 20, 20]        [64, 20, 20]        3022727
(70)    conv_linear                   [64, 20, 20]        [64, 20, 20]        3026887
(71)    route: 70, 66                 -                   [93, 20, 20]        -
(72)    shuffle                       [93, 20, 20]        [93, 400]           -
(73)    route: 52, 62, 72             -                   [93, 8400]          -
(74)    detect_v8                     [93, 8400]          [8400, 33]          3052087

Number of unused weights left: 18446744073709540969
deepstream-app: yolo.cpp:437: NvDsInferStatus Yolo::buildYoloNetwork(std::vector<float>&, nvinfer1::INetworkDefinition&): Assertion `0' failed.

**Comment by jiangdizhao** [comment]
[v17.zip](https://github.com/marcoslucianops/DeepStream-Yolo/files/10636813/v17.zip)

**Comment by jiangdizhao** [comment]
This is a yolov8n model, with 29 classes of objects, Would you please convert it on your computer and let me know your results ? thank you @marcoslucianops

**Comment by marcoslucianops** [comment]
I did a new update in the repo to fix the problem. Please, convert your model again using the new gen_wts_yoloV8.py file and test it.

**Comment by jiangdizhao** [comment]
Perfect, problem solved! Thanks a lot !
